Add `apply_mask` function to `mask` sub-package

b-reyes opened the pull request:
This PR implements the `apply_mask` function discussed in https://github.com/OSOceanAcoustics/echopype/pull/817#issuecomment-1351934933 and subsequent comments in #817.

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+
+    Parameters
+    ----------
+    mask: xr.DataArray or str or pathlib.Path or list of xr.DataArray or str or pathlib.Path
+        The mask(s) to apply to the variable specified by ``var_name``. This input can be a
+        single input or list that corresponds to a DataArray or a path to a DataArray. If a path
+        is provided this should point to a zarr or netcdf file with only one array in it.
```
```suggestion
    mask: xr.DataArray, str, pathlib.Path, or a list of these datatypes
        The mask(s) to be applied. 
        Can be a single input or list that corresponds to a DataArray or a path. If a path
        is provided this should point to a zarr or netcdf file with only one data variable in it.
```
Changing the wording to reflect the scope of this function and clarify terminology.

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+
+    Raises
+    ------
+    RuntimeError
```
```suggestion
    ValueError
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+    else:
+
+        if not isinstance(storage_options_mask, dict):
+            raise RuntimeError(
```
```suggestion
            raise ValueError(
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+
+def apply_mask(
+    source_ds: Union[xr.Dataset, str, pathlib.Path],
+    var_name: str,
```
I think we should default this to `Sv`.

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+    ------
+    TypeError
+        If ``var_name`` or ``fill_value`` are not an accepted type
+    RuntimeError
+        If the Dataset ``source_ds`` does not contain ``var_name``
+    RuntimeError
+        If ``fill_value`` is an array and not the same shape as ``var_name``
```
Please do overarching changes of `RuntimeError` to `ValueError` in this PR.

We'll worry about existing ones in other recent PRs later.

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+        Points to a Dataset that contains the variable the mask should be applied to
+    var_name: str
+        The variable name in ``source_ds`` that the mask should be applied to
+    mask: xr.DataArray or str or pathlib.Path or list of xr.DataArray or str or pathlib.Path
+        The mask(s) to apply to the variable specified by ``var_name``. This input can be a
+        single input or list that corresponds to a DataArray or a path to a DataArray. If a path
+        is provided this should point to a zarr or netcdf file with only one array in it.
```
```suggestion
    mask: xr.DataArray, str, pathlib.Path, or a list of these datatypes
        The mask(s) to be applied. 
        Can be a single input or list that corresponds to a DataArray or a path. If a path
        is provided this should point to a zarr or netcdf file with only one data variable in it.
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+
+    Parameters
+    ----------
+    source_ds: xr.Dataset or str or pathlib.Path
```
```suggestion
    source_ds: xr.Dataset, str, or pathlib.Path
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+        single input or list that corresponds to a DataArray or a path to a DataArray. If a path
+        is provided this should point to a zarr or netcdf file with only one array in it.
+    fill_value: int or float or np.ndarray or xr.DataArray, default=np.nan
+        Specifies the value(s) at false indices
```
```suggestion
        Value(s) at masked indices
```

leewujung commented on echopype/tests/mask/test_mask.py:
```diff
@@ -219,6 +275,133 @@ def test_frequency_differencing(n: int, n_chan_freq: int,
     assert out.name == "mask"
 
 
+@pytest.mark.parametrize(
+    ("n", "n_chan", "var_name", "mask", "mask_file", "fill_value", "is_delayed", "var_masked_truth"),
+    [
+        (2, 1, "var1", np.identity(2), None, np.nan, False, np.array([[1, np.nan], [np.nan, 1]])),
+        (2, 1, "var1", np.identity(2), None, 2.0, False, np.array([[1, 2.0], [2.0, 1]])),
+        (2, 1, "var1", np.identity(2), None, np.array([[[np.nan, np.nan], [np.nan, np.nan]]]),
+         False, np.array([[1, np.nan], [np.nan, 1]])),
+        (2, 1, "var1", np.identity(2), None, xr.DataArray(data=np.array([[[np.nan, np.nan], [np.nan, np.nan]]]),
+                                                          coords={"channel": ["chan1"],
+                                                                  "ping_time": [0, 1],
+                                                                  "range_sample": [0, 1]}),
+         False, np.array([[1, np.nan], [np.nan, 1]])),
+        (2, 1, "var1", [np.identity(2), np.array([[0, 1], [0, 1]])], [None, None], 2.0,
+         False, np.array([[2.0, 2.0], [2.0, 1]])),
+        (2, 1, "var1", np.identity(2), None, 2.0, True, np.array([[1, 2.0], [2.0, 1]])),
+        (2, 1, "var1", np.identity(2), "test.zarr", 2.0, True, np.array([[1, 2.0], [2.0, 1]])),
+        (2, 1, "var1", [np.identity(2), np.array([[0, 1], [0, 1]])], ["test0.zarr", "test1.zarr"], 2.0,
+         False, np.array([[2.0, 2.0], [2.0, 1]])),
+        (2, 1, "var1", [np.identity(2), np.array([[0, 1], [0, 1]])], ["test0.zarr", None], 2.0,
+         False, np.array([[2.0, 2.0], [2.0, 1]])),
+    ],
+    ids=["single_mask_default_fill", "single_mask_float_fill", "single_mask_np_array_fill",
+         "single_mask_DataArray_fill", "list_mask_all_np", "single_mask_ds_delayed",
+         "single_mask_as_path", "list_mask_all_path", "list_mask_some_path"]
+)
+def test_apply_mask(n: int, n_chan: int, var_name: str,
+                    mask: Union[np.ndarray, List[np.ndarray]],
+                    mask_file: Optional[Union[str, List[str]]],
+                    fill_value: Union[int, float, np.ndarray, xr.DataArray],
+                    is_delayed: bool, var_masked_truth: np.ndarray):
```
I think what test contains right now is more of the integrated test flavor, and some of those can be split out to unit test the mask validation and collection function.

My suggestion is create/split things into the following:
- unit tests for `validate_and_collect_mask_input`: 
  - test allowable types of mask sources
  - test the allowable forms of mask (list or single)
- unit tests for `_check_var_name_fill_value`
  - test whether `var_name` is in `source_ds`
  - test whether `fill_value` is of the same shape as `source_ds["var_name"]`
- integration tests that actually applies the tests (you have these already)

leewujung commented on echopype/utils/io.py:
```diff
@@ -318,29 +318,29 @@ def env_indep_joinpath(*args: Tuple[str, ...]) -> str:
-    source_ds: xr.Dataset or str or pathlib.Path
-        A source that points to a Dataset. If the input is a path, it specifies
-        the path to a zarr or netcdf file.
+    source_ds_da: xr.Dataset or xr.DataArray or str or pathlib.Path
```
```suggestion
    source_ds_da: xr.Dataset, xr.DataArray, str or pathlib.Path
```

leewujung reviewed (changes requested):
@b-reyes : thanks for the PR! All of my actual comments are textual, with the exception of the tests -- I think it would be good to set up unit tests for the functions used by the main function. 

**Update:** I think we should consider refactoring the test as a separate PR as part of the test hackday, so that this can be merged by end of day for @emiliom to add the provenance piece.

b-reyes commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+            if isinstance(mask_val, str):
+                # open up DataArray using mask path
+                mask[mask_ind] = xr.open_dataarray(
+                    mask_val, engine=file_type, chunks="auto", **storage_options_mask[mask_ind]
```
While working on another PR I found that `chunks="auto"` will fail for netcdf. This should be replaced with `chunks={}`.

b-reyes commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+
+        if isinstance(mask, str):
+            # open up DataArray using mask path
+            mask = xr.open_dataarray(mask, engine=file_type, chunks="auto", **storage_options_mask)
```
While working on another PR I found that `chunks="auto"` will fail for netcdf. This should be replaced with `chunks={}`.

b-reyes commented on echopype/mask/api.py:
```diff
@@ -17,6 +17,231 @@
+
+    if isinstance(source_ds, str):
+        # open up Dataset using source_ds path
+        source_ds = xr.open_dataset(
```
While working on another PR I found that `chunks="auto"` will fail for netcdf. This should be replaced with `chunks={}`.

b-reyes commented:
@leewujung I have addressed all of your comments. As you suggested, we will wait to do the test refactoring as a separate PR as part of the test hackday.

leewujung reviewed (approved):
@b-reyes : thank you! I will merge this now.
